src/conversion.cc

```diff
@@ -0,0 +1,199 @@
+  if (available_capacity <= 0) return ports;
+
+  // Create request portfolio
+  RequestPortfolio<Material>::Ptr port(new RequestPortfolio<Material>());
+
+  // Create material request
+  Material::Ptr mat;
+  if (inrecipe_name.empty()) {
+    mat = cyclus::NewBlankMaterial(available_capacity);
+  } else {
+    mat = Material::CreateUntracked(available_capacity, context()->GetRecipe(inrecipe_name));
+  }
+
+  // Add request for all commodities using default preference
+  for (std::vector<std::string>::iterator it = incommods.begin();
+       it != incommods.end(); ++it) {
+    Request<Material>* req = port->AddRequest(mat, this, *it);
+  }
+
+  // Add capacity constraint to ensure we never get more feed than capacity
+  CapacityConstraint<Material> cc(available_capacity);
+  port->AddConstraint(cc);
+
+  ports.insert(port);
+  return ports;
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+std::set<BidPortfolio<Material>::Ptr> Conversion::GetMatlBids(
+  CommodMap<Material>::type& commod_requests) {
+  std::set<BidPortfolio<Material>::Ptr> ports;
+
+  // Check if we have material to offer
+  if (output.quantity() <= 0) return ports;
+
+  // Create bid portfolio
+  BidPortfolio<Material>::Ptr port(new BidPortfolio<Material>());
+
+  // Respond to requests for our commodity
+  std::vector<Request<Material>*>& requests = commod_requests[outcommod];
+  for (std::vector<Request<Material>*>::iterator it = requests.begin();
+      it != requests.end(); ++it) {
+
+    double available = output.quantity();
+    double requested = (*it)->target()->quantity();
+    double offer_qty = std::min(available, requested);
+
+    if (offer_qty > 0) {
+      Material::Ptr offer = Material::CreateUntracked(offer_qty, output.Peek()->comp());
+      port->AddBid(*it, offer, this);  // Note: *it, not **it
+    }
+  }
+
+  // Add capacity constraint so we never give out more than we have
+  CapacityConstraint<Material> cc(output.quantity());
+  port->AddConstraint(cc);
+
+  ports.insert(port);
+  return ports;
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+void Conversion::AcceptMatlTrades(
+  const std::vector<std::pair<Trade<Material>, Material::Ptr>>& responses) {
+
+  for (std::vector<std::pair<Trade<Material>, Material::Ptr>>::const_iterator it =
+      responses.begin(); it != responses.end(); ++it) {
+
+    Material::Ptr mat = it->second;
+    cyclus::toolkit::MatQuery mq(mat);
+
+    // Add material to the input buffer
+    input.Push(mat);
+    
+  }
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+void Conversion::GetMatlTrades(
+  const std::vector<Trade<Material>>& trades,
+  std::vector<std::pair<Trade<Material>, Material::Ptr>>& responses) {
+
+  for (std::vector<Trade<Material>>::const_iterator it = trades.begin();
+      it != trades.end(); ++it) {
+
+    double requested_qty = it->amt;
+    Material::Ptr response = output.Pop(requested_qty);
```

Comment:
No need for an extra variable here
```suggestion
    Material::Ptr response = output.Pop(it->amt);
```

Reply:
I was wondering if we used the explicit pattern somewhere else (and I was correcting you on something that you just copied from elsewhere).  For a tutorial, it may be more valuable to include an explicit named variable here for clarity, but I think we can leave it out.  It's kind of a judgment call, and I probably come down on different sides of this in different cases depending on how much clarity I think is needed/provided by this pattern.